File: lab10database.py
import sqlite3
import threading

class DatabaseOperations:
    def __init__(self):
        self.connection = sqlite3.connect("example.db")
        self.cursor = self.connection.cursor()
    
    def create_table(self):
        try:
            self.cursor.execute('''CREATE TABLE IF NOT EXISTS users (id INTEGER PRIMARY KEY, name TEXT)''')
            self.connection.commit()
        except Exception as e:
            print(f"Error: {e}")

    def insert_data(self, name):
        try:
            self.cursor.execute('''INSERT INTO users (name) VALUES (?)''', (name,))
            self.connection.commit()
        except Exception as e:
            print(f"Error: {e}")

    def fetch_data(self):
        try:
            self.cursor.execute('''SELECT * FROM users''')
            return self.cursor.fetchall()
        except Exception as e:
            print(f"Error: {e}")

    def close(self):
        self.connection.close()

class WorkerThread(threading.Thread):
    def __init__(self, db_operations):
        threading.Thread.__init__(self)
        self.db_operations = db_operations
    
    def run(self):
        self.db_operations.insert_data("John Doe")
        self.db_operations.insert_data("Jane Smith")
        print(self.db_operations.fetch_data())

def main():
    db_operations = DatabaseOperations()
    db_operations.create_table()
    
    thread1 = WorkerThread(db_operations)
    thread2 = WorkerThread(db_operations)
    
    thread1.start()
    thread2.start()
    
    thread1.join()
    thread2.join()

    db_operations.close()

if __name__ == "__main__":
    main()

File: lab10database.java
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Statement;

class DatabaseOperations {
    private Connection connection;
    
    public DatabaseOperations() {
        try {
            connection = DriverManager.getConnection("jdbc:sqlite:example.db");
        } catch (SQLException e) {
            System.out.println("Error: " + e.getMessage());
        }
    }

    public void createTable() {
        try {
            Statement stmt = connection.createStatement();
            stmt.execute("CREATE TABLE IF NOT EXISTS users (id INTEGER PRIMARY KEY, name TEXT)");
        } catch (SQLException e) {
            System.out.println("Error: " + e.getMessage());
        }
    }

    public void insertData(String name) {
        try {
            PreparedStatement pstmt = connection.prepareStatement("INSERT INTO users (name) VALUES (?)");
            pstmt.setString(1, name);
            pstmt.executeUpdate();
        } catch (SQLException e) {
            System.out.println("Error: " + e.getMessage());
        }
    }

    public void fetchData() {
        try {
            Statement stmt = connection.createStatement();
            ResultSet rs = stmt.executeQuery("SELECT * FROM users");
            while (rs.next()) {
                System.out.println(rs.getInt("id") + ": " + rs.getString("name"));
            }
        } catch (SQLException e) {
            System.out.println("Error: " + e.getMessage());
        }
    }

    public void close() {
        try {
            connection.close();
        } catch (SQLException e) {
            System.out.println("Error: " + e.getMessage());
        }
    }
}

class WorkerThread extends Thread {
    private DatabaseOperations dbOperations;
    
    public WorkerThread(DatabaseOperations dbOperations) {
        this.dbOperations = dbOperations;
    }
    
    @Override
    public void run() {
        dbOperations.insertData("John Doe");
        dbOperations.insertData("Jane Smith");
        dbOperations.fetchData();
    }
}

public class lab10database {
    public static void main(String[] args) {
        DatabaseOperations dbOperations = new DatabaseOperations();
        dbOperations.createTable();
        
        WorkerThread thread1 = new WorkerThread(dbOperations);
        WorkerThread thread2 = new WorkerThread(dbOperations);
        
        thread1.start();
        thread2.start();
        
        try {
            thread1.join();
            thread2.join();
        } catch (InterruptedException e) {
            System.out.println("Error: " + e.getMessage());
        }
        
        dbOperations.close();
    }
}
